onnxruntime/core/providers/cuda/nn/conv_transpose.cc

```diff
@@ -155,42 +157,49 @@ Status ConvTranspose<T>::DoConvTranspose(OpKernelContext* context, bool dynamic_
       s_.algo = perf.algo;
       s_.workspace_bytes = perf.memory;
     }
-  }
 
-  if (!y_data) {
-    auto y_dims = s_.y_dims;
-    if (x_dimensions == 3) {
-      y_dims.erase(y_dims.begin() + 2);
+    // The following block will be executed in case there has been no change in the shapes of the
+    // input and the filter compared to the previous run
+    if (!y_data) {
+      auto y_dims = s_.y_dims;
+      if (x_dimensions == 3) {
+        y_dims.erase(y_dims.begin() + 2);
+      }
+      Tensor* Y = context->Output(0, TensorShape(y_dims));
+      y_data = reinterpret_cast<CudaT*>(Y->template MutableData<T>());
+
+      // Bail out early if one of the output dimensions is zero.
+      if (Y->Shape().Size() == 0) {
+        return Status::OK();
+      }
     }
-    Tensor* Y = context->Output(0, TensorShape(y_dims));
-    y_data = reinterpret_cast<CudaT*>(Y->template MutableData<T>());
-  }
 
-  const auto alpha = Consts<CudaT>::One;
-  const auto beta = Consts<CudaT>::Zero;
-
-  IAllocatorUniquePtr<void> workspace = GetScratchBuffer<void>(s_.workspace_bytes);
-
-  CUDNN_RETURN_IF_ERROR(
-      cudnnConvolutionBackwardData(
-          CudnnHandle(),
-          &alpha,
-          s_.filter_desc,
-          w_data,
-          s_.x_tensor,
-          x_data,
-          s_.conv_desc,
-          s_.algo,
-          workspace.get(),
-          s_.workspace_bytes,
-          &beta,
-          s_.y_tensor,
-          y_data));
-
-  if (has_bias) {
-    const Tensor* B = dynamic_padding ? context->Input<Tensor>(3) : context->Input<Tensor>(2);
-    auto b_data = reinterpret_cast<const CudaT*>(B->template Data<T>());
-    CUDNN_RETURN_IF_ERROR(cudnnAddTensor(CudnnHandle(), &alpha, s_.b_tensor, b_data, &alpha, s_.y_tensor, y_data));
+    const auto alpha = Consts<CudaT>::One;
+    const auto beta = Consts<CudaT>::Zero;
+
+    IAllocatorUniquePtr<void> workspace = GetScratchBuffer<void>(s_.workspace_bytes);
+
+    CUDNN_RETURN_IF_ERROR(
+        cudnnConvolutionBackwardData(
+            CudnnHandle(),
+            &alpha,
+            s_.filter_desc,
+            w_data,
+            s_.x_tensor,
+            x_data,
+            s_.conv_desc,
+            s_.algo,
+            workspace.get(),
+            s_.workspace_bytes,
+            &beta,
+            s_.y_tensor,
+            y_data));
+
+    if (has_bias) {
+      const Tensor* B = dynamic_padding ? context->Input<Tensor>(3) : context->Input<Tensor>(2);
+      auto b_data = reinterpret_cast<const CudaT*>(B->template Data<T>());
+      CUDNN_RETURN_IF_ERROR(cudnnAddTensor(CudnnHandle(), &alpha, s_.b_tensor, b_data, &alpha, s_.y_tensor, y_data));
+    }
   }
```

Comment:
Move this brace down thus extending holding onto the mutex after all CuDNN calls